src/error_handling_helpers.h

```diff
@@ -0,0 +1,160 @@
+// Copyright 2014 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Note: migrated from rmw/error_handling.h in 2017-04
+
+#ifndef ERROR_HANDLING_HELPERS_H_
+#define ERROR_HANDLING_HELPERS_H_
+
+// When this define evaluates to true (default), then messages will printed to
+// stderr when an error is encountered while setting the error state.
+// For example, when memory cannot be allocated or a previous error state is
+// being overwritten.
+#ifndef RCUTILS_REPORT_ERROR_HANDLING_ERRORS
+#define RCUTILS_REPORT_ERROR_HANDLING_ERRORS 1
+#endif
+
+// When this define evaluates to true (default), then a warning will be written
+// to stderr using RCUTILS_SAFE_FWRITE_TO_STDERR.
+#ifndef RCUTILS_WARN_ON_TRUNCATION
+#define RCUTILS_WARN_ON_TRUNCATION 1
+#endif
+
+#define __STDC_WANT_LIB_EXT1__ 1  // indicate we would like memmove_s if available
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <rcutils/error_handling.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+size_t
+__rcutils_copy_string(char * dst, size_t dst_size, const char * src)
+{
+  // doesn't matter how long src actually is if it is longer than dst, so limit to dst + 1
```

Comment:
I see.  And now I actually understand a bit more about why this feels "weird" to me; it's because these helpers are all in a header file, which is intrinsically "public" to me in C.  Further, I just realized that if more than one file did a `#include "./error_handling_helpers.h"`, the link would fail because of a multiple-symbol issue.  I think I'd feel a lot better about this if this header file just had the prototype for `__rcutils_format_error_string` in it, and then all of the implementation (including the helpers) was in a C file.  This would also solve the multi-symbol issue (though that is less important).  What do you think about that approach?

Reply:
I could do that, but the reason it's in a header file is so that I can create a test for it but I was explicitly not including it in more than one place.

It might feel public, but I don't consider it public since I don't install the header (also why it's in `src` and not in `include`). I was also hoping that it would get inlined because it was in the header, and avoiding another translation unit. That being said I don't mind moving the implementation into a `.c` file.

But does that address the concerns about these functions being unsafe to use if you ignore the assumptions?

Reply:
> But does that address the concerns about these functions being unsafe to use if you ignore the assumptions?

Yeah, if they are hidden in a C file, then there is no way for any external users to call them, so I'm much more comfortable with eliding internal sanity checks.

Reply:
But the prototypes are still in a header, so it doesn't change their visibility as it is right now, right? Or do you mean that I should just rename the file to be `.c` and include that (because that also works)?

Reply:
@clalancette can you have a look at that commit and see if there's more you want me to do (I'm happy to continue iterating on it)

Reply:
I'm going to push to get this merged. I'm happy to follow up with more changes in an additional PR if you have more feedback later @clalancette.